# UniOP: self-supervised universal operon prediction for prokaryotic (meta-)genomics data
The manuscript is currently being prepared.

## Overview
We present **UniOP**, a fast and accurate computational method for operon prediction, independent of experimental information or functional information. UniOP takes as input one or more prokaryotic genomes or metagenomic assembled genomes (MAGs), utilizing the features extracted solely from genomic sequence. These features include:
(i) transcription co-directionality;
(ii) intergenic distances; 
(iii) conserved gene neighborhood. 
**UniOP** is composed of two components: an intergenic distance-based module **UniOP_dst** and a conservation-based module **UniOP_cons**. **UniOP_cons** is a regression model learning to predict pseudo labels generated by **UniOP_dst** using conservation information, which is identified by [Spacedust](https://github.com/soedinglab/spacedust). The predictions from the **UniOP_dst** and **UniOP_cons** are combined to generate the final operon prediction. 


## Installation
### Prerequisites
**UniOP** requires:

* Python3
* Tensorflow
* Prodigal [https://github.com/hyattpd/Prodigal/wiki/installation]
* spacedust [https://github.com/soedinglab/spacedust]
* mash [https://github.com/marbl/mash]


## Input
The starting point is either a FASTA file of the nucleotide genome sequence (`.fna`) or protein-coding sequences (**CDS**) (`.faa`)(typically by running a gene prediction program like [Prodigal](https://github.com/hyattpd/Prodigal))
