# UniOP: self-supervised universal operon prediction for prokaryotic (meta-)genomics data
The manuscript is currently being prepared.

## Overview
We present **UniOP**, a fast and accurate computational method for operon prediction, independent of experimental or functional information. UniOP takes as input prokaryotic genomes or metagenomic assembled genomes (MAGs), utilizing the features extracted solely from genomic sequence. These features include:
(i) transcription co-directionality;
(ii) intergenic distances; 
(iii) conserved gene neighborhood. 
**UniOP** is composed of two components: an intergenic distance-based module **UniOP_dst** and a conservation-based module **UniOP_cons**. **UniOP_cons** is a regression model learning to predict pseudo labels generated by **UniOP_dst** using conservation information identified by [Spacedust](https://github.com/soedinglab/spacedust). The predictions from the **UniOP_dst** and **UniOP_cons** are combined to generate the final operon prediction. 


## Installation
### Dependencies
**UniOP** requires:
UniOP requires the installation of [Prodigal](https://github.com/hyattpd/Prodigal/wiki/installation), [spacedust](https://github.com/soedinglab/spacedust), [mash](https://github.com/marbl/mash) in the working directory.
```
* Python (>=3.10.12)
* Prodigal
* spacedust 
* mash
```
### Install packages
```
bash environment.sh
```
### Install UniOP
```
wget https://github.com/hongsua/UniOP/archive/refs/heads/main.zip
unzip main.zip
```
Install **Prodigal**,**spacedust**, and **mash** into the working directory, ~/UniOP-main/src.
```
cd UniOP-main/src
(The details about how to install the above dependencies.)
```
## Usage
## Input
The starting point is either a FASTA file of the nucleotide genome sequence (`.fna`) or protein-coding sequences (**CDS**) (`.faa`)(typically by running a gene prediction program like [Prodigal](https://github.com/hyattpd/Prodigal))

### Operon prediction based on intergenic distance only
```
python3 UniOP_dst.py -i ../demo/GCF_000005845.2.fna
```
This will output gene predictions: **GCF_000005845.2.faa**, **GCF_000005845.2.gff**, and the operon prediction: **dist.pred**.
### Operon prediction based on intergenic distance and conserved adjacent gene pairs
```
python3 UniOP.py -a ../demo/GCF_000005845.2.faa -db_msh ../data/ncbi_reference.msh -db_pred ../data/dist_prediction
```
This will output a directory: **spacedust** containing the gene clusters detected by the program **spacedust**, and the final prediction: **uniop.pred**.

## Support
If you have questions or found any bug in the program, please write to us at
hong.su[at]mpinat.mpg.de
