# UniOP: self-supervised universal operon prediction for prokaryotic (meta-)genomics data
The manuscript is in preparation.

## Overview
We present UniOP, a fast and accurate computational method for operon prediction, independent of experimental information or functional information. UniOP takes as input one or more prokaryotic genomes or metagenomic assembled genomes (MAGs), utilizing the features extracted solely from genomic sequence. These features include:
(i) transcription co-directionality, where genes transcribed in opposite directions are part of different operons;
(ii) intergenic distances, as genes within the same operon tend to have smaller intergenic distances compared to genes not belonging to the same operon; 
(iii) conserved gene neighborhood, where genes within an operon exhibit conserved gene order and strand orientation across genomes due to evolutionary relationships or horizontal gene transfer. 
We estimate the probability of adjacent gene pairs belonging to the same operon based on the intergenic distances (denoted by UniOP_dst) and employ [Spacedust](https://github.com/soedinglab/spacedust) for identifying conserved gene clusters in reference genomes. We adopt a self-training approach where a regression model (denoted by UniOP_cons) learns to predict pseudo labels generated by the UniOP_dst using conservation information. The predictions from the UniOP_dst and UniOP_cons are combined to generate the final operon prediction. 


## Installation
